### ISSUE TITLE
pytorch core test failure: AssertionError: bool indices are not supported yet

### ISSUE THREAD

**Issue by desertfire** [pytorch/torchdynamo#1267]
Repro:
```
PYTORCH_TEST_WITH_INDUCTOR=1   pytest test/test_indexing.py -k test_empty_index_cuda
```
Error:
```
Traceback (most recent call last):
  File "/fsx/users/binbao/pytorch/test/test_indexing.py", line 962, in test_empty_index
    def test_empty_index(self, device):
  File "/fsx/users/binbao/pytorch/test/test_indexing.py", line 965, in <graph break in test_empty_index>
    self.assertEqual(x[idx].numel(), 0)
  File "/fsx/users/binbao/pytorch/test/test_indexing.py", line 970, in <graph break in test_empty_index>
    self.assertEqual(x, y)
  File "/fsx/users/binbao/torchdynamo-tip/torchdynamo/eval_frame.py", line 167, in _fn
    return fn(*args, **kwargs)
  File "/fsx/users/binbao/pytorch/functorch/_src/aot_autograd.py", line 904, in forward
    return compiled_f(
  File "/fsx/users/binbao/pytorch/functorch/_src/aot_autograd.py", line 890, in new_func
    compiled_fn = create_aot_dispatcher_function(
  File "/fsx/users/binbao/torchdynamo-tip/torchdynamo/utils.py", line 76, in time_wrapper
    r = func(*args, **kwargs)
  File "/fsx/users/binbao/pytorch/functorch/_src/aot_autograd.py", line 468, in create_aot_dispatcher_function
    return aot_dispatch_base(flat_fn, fake_flat_tensor_args, aot_config)
  File "/fsx/users/binbao/pytorch/functorch/_src/aot_autograd.py", line 278, in aot_dispatch_base
    compiled_fw = aot_config.fw_compiler(fw_module, flat_args)
  File "/fsx/users/binbao/pytorch/functorch/_src/aot_autograd.py", line 237, in f
    out_f = compiler(fx_g, inps)
  File "/fsx/users/binbao/torchdynamo-tip/torchdynamo/utils.py", line 76, in time_wrapper
    r = func(*args, **kwargs)
  File "/fsx/users/binbao/torchdynamo-tip/torchinductor/compile_fx.py", line 246, in fw_compiler
    return compile_fx_inner(
  File "/fsx/users/binbao/torchdynamo-tip/torchdynamo/debug_utils.py", line 332, in debug_wrapper
    compiled_fn = compiler(gm, example_inputs, **kwargs)
  File "/fsx/users/binbao/torchdynamo-tip/torchinductor/debug.py", line 182, in inner
    return fn(*args, **kwargs)
  File "/fsx/users/binbao/conda/envs/release/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/fsx/users/binbao/torchdynamo-tip/torchinductor/compile_fx.py", line 83, in compile_fx_inner
    wrap(graph.run)(*example_inputs)
  File "/fsx/users/binbao/torchdynamo-tip/torchdynamo/utils.py", line 76, in time_wrapper
    r = func(*args, **kwargs)
  File "/fsx/users/binbao/torchdynamo-tip/torchinductor/graph.py", line 143, in run
    return super().run(*args)
  File "/fsx/users/binbao/pytorch/torch/fx/interpreter.py", line 129, in run
    self.env[node] = self.run_node(node)
  File "/fsx/users/binbao/torchdynamo-tip/torchinductor/graph.py", line 303, in run_node
    result = super().run_node(n)
  File "/fsx/users/binbao/pytorch/torch/fx/interpreter.py", line 162, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/fsx/users/binbao/torchdynamo-tip/torchinductor/graph.py", line 255, in call_function
    raise LoweringException(e, target, args, kwargs) from e
torchinductor.exc.LoweringException: AssertionError: bool indices are not supported yet
  target: aten.index_put_.default
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='arg0_1', layout=FixedLayout('cuda', torch.int64, size=[s0, s1], stride=[s1, 1]))
  ))
  args[1]: [TensorBox(StorageBox(
    Pointwise(
      'cuda',
      torch.bool,
      to_dtype(constant(0, torch.float32), torch.bool),
      ranges=[4, 3],
      origins={_to_copy}
    )
  ))]
  args[2]: TensorBox(StorageBox(
    Pointwise(
      'cpu',
      torch.int64,
      constant(-1, torch.int64),
      ranges=[],
      origins={lift_fresh_copy}
    )
  ))
```

**Comment by desertfire** [comment]
@sangongs , feel free to ask @ngimel or me if you have any questions.

**Comment by sangongs** [comment]
I tried to fallback when there is a bool index:
```
@register_lowering(aten.index_put_, type_promote=False)
def index_put_(self, indices, values, accumulate=False):
    for index in indices:
        if index.get_dtype() == torch.bool:
            return TensorBox.create(ir.FallbackKernel.create(aten.index_put_, self, indices, values, accumulate))
    ...
```

But got a CUDA memory leak:
```
RuntimeError: CUDA driver API confirmed a leak in test_indexing.TestIndexingCUDA.test_empty_index_cuda! Caching allocator allocated memory was 0 and is now reported as 3584 on device 0. CUDA driver allocated memory was 2833121280 and is now 2841509888.
```

The test passes if I set `PYTORCH_TEST_SKIP_CUDA_MEM_LEAK_CHECK=1`.

Questions:
- Is fallback a correct way to fix this?
- Any pointer for the CUDA memory leak?

**Comment by ngimel** [comment]
Fallback is ok for now, going forward we will have to do the same thing eager does - convert bool mask(s) to long indices via nonzero, and call `index` with long indices. There are a lot of tests with fallbacks currently causing mem leaks (see #1367), so skipping it for now is ok I think.

**Comment by desertfire** [comment]
I actually think we are not in a hurry to fix this unit test with a fallback given it's not blocking any benchmark model. We can always add the test to the skip list if CI is a concern. It will be more helpful if @sangongs can implement the actual op lowering.

**Comment by ngimel** [comment]
The code pointer in the eager where bool masks are converted to long indices https://github.com/pytorch/pytorch/blob/7b3a0ff87a91c98574ddfd98532062d78c49b2c7/aten/src/ATen/native/IndexingUtils.h#L17-L49

**Comment by sangongs** [comment]
> The code pointer in the eager where bool masks are converted to long indices https://github.com/pytorch/pytorch/blob/7b3a0ff87a91c98574ddfd98532062d78c49b2c7/aten/src/ATen/native/IndexingUtils.h#L17-L49 Also, note that this requires `nonzero` op that currently doesn't have lowering/decomposition, and you can't create a fallback for it because fallback hard-codes the output shape, whereas the output shape for `nonzero` is data dependent, so I'm not sure how to proceed there.

I have the same doubt. Currently, `torch.nonzero` causes a graph break even with `torchdynamo.config.dynamic_shapes = True`. So I guess nonzero is hard to lower.

**Comment by sangongs** [comment]
Looks like although we cannot lower `nonzero` or `aten.index(bool)`, we could lower `aten.index_put_(bool)`, which is actually a pointwise operation.

**Comment by ngimel** [comment]
Not really, if your `src` for index_put is also a tensor - that still requires converting bool indices to long. If the src for index_put is a number/0d tensor, and there are only bool indices, not a mixture of bool and long indices, then yeah, that's possible, that's essentially a masked_fill operation, and eager implementation shows the necessary checks for dispatching index_put to masked_fill.

**Comment by sangongs** [comment]
I got what you mean. Do you think it is necessary to lower the special case that is essentially masked_fill? @desertfire

**Comment by ngimel** [comment]
It's pretty common and perf sensitive (that's why we have special casing in eager), so yes.

**Comment by sangongs** [comment]
Fixed by PR #1419. Regression tested on commit 2e883d4655ce4ad85b1a2af5cf9908c0032549c5.
